mmdet/models/detectors/deformable_detr.py

```diff
@@ -1,12 +1,575 @@
+            query = query.unsqueeze(0).expand(batch_size, -1, -1)
+            reference_points = self.reference_points_fc(query_pos).sigmoid()
+
+        # [] -> []  # TODO
+        query = query.permute(1, 0, 2)
+        memory = memory.permute(1, 0, 2)
+        query_pos = query_pos.permute(1, 0, 2)
+
+        decoder_inputs_dict = dict(
+            query=query,
+            query_pos=query_pos,
+            memory=memory,
+            reference_points=reference_points)
+        head_inputs_dict = dict(
+            enc_outputs_class=enc_outputs_class if self.as_two_stage else None,
+            enc_outputs_coord=enc_outputs_coord_unact.sigmoid()
+            if self.as_two_stage else None)  # TODO: Is this right ?
+        return decoder_inputs_dict, head_inputs_dict
+
+    def forward_decoder(self, memory, memory_mask, query, query_pos,
+                        reference_points, spatial_shapes, level_start_index,
+                        valid_ratios) -> Dict:  # TODO: typehint
+        # TODO: Doc
+        inter_states, inter_references = self.decoder(
+            query=query,
+            key=None,
+            value=memory,
+            query_pos=query_pos,
+            key_padding_mask=memory_mask,
+            reference_points=reference_points,
+            spatial_shapes=spatial_shapes,
+            level_start_index=level_start_index,
+            valid_ratios=valid_ratios,
+            reg_branches=self.bbox_head.reg_branches
+            if self.with_box_refine else None)
+        head_inputs_dict = dict(
+            hidden_states=inter_states,
+            init_reference=reference_points,
+            inter_references=inter_references)
+        return head_inputs_dict
+
+    def gen_encoder_output_proposals(
+            self, memory: Tensor, memory_padding_mask: Tensor,
+            spatial_shapes: Tensor) -> Tuple[Tensor, Tensor]:
+        """Generate proposals from encoded memory.
+
+        Args:
+            memory (Tensor) : The output of encoder,
+                has shape (batch_size, num_key, embed_dim).  num_key is
+                equal the number of points on feature map from
+                all level.
+            memory_padding_mask (Tensor): Padding mask for memory.
+                has shape (batch_size, num_key).
+            spatial_shapes (Tensor): The shape of all feature maps.
+                has shape (num_level, 2).
+
+        Returns:
+            tuple: A tuple of feature map and bbox prediction.
+
+                - output_memory (Tensor): The input of decoder,  \
+                    has shape (batch_size, num_key, embed_dim).  num_key is \
+                    equal the number of points on feature map from \
+                    all levels.
+                - output_proposals (Tensor): The normalized proposal \
+                    after a inverse sigmoid, has shape \
+                    (batch_size, num_keys, 4).
+        """
+
+        N, S, C = memory.shape
```

Comment:
Please change to a name that more clear, such as  batch_size, num_key, embed_dim